File: java/norito_java/src/main/java/org/hyperledger/iroha/norito/NoritoDecoder.java
// Copyright 2024 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.norito;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

/** Reader for Norito payloads. */
public final class NoritoDecoder {
  private final ByteBuffer buffer;
  private final int flags;
  private final int flagsHint;

  public NoritoDecoder(byte[] data, int flags) {
    this(data, flags, NoritoHeader.MINOR_VERSION);
  }

  public NoritoDecoder(byte[] data, int flags, int flagsHint) {
    this(ByteBuffer.wrap(Arrays.copyOf(data, data.length)), flags, flagsHint);
  }

  public NoritoDecoder(ByteBuffer data, int flags, int flagsHint) {
    this.buffer = data.slice().order(ByteOrder.LITTLE_ENDIAN);
    this.flags = flags;
    this.flagsHint = flagsHint;
  }

  public int flags() {
    return flags;
  }

  public int flagsHint() {
    return flagsHint;
  }

  public boolean compactLenActive() {
    return explicitlyEnabled(NoritoHeader.COMPACT_LEN);
  }

  private boolean explicitlyEnabled(int flag) {
    return (flags & flag) != 0;
  }

  public int remaining() {
    return buffer.remaining();
  }

  public byte[] readBytes(int count) {
    ensure(count);
    byte[] out = new byte[count];
    buffer.get(out);
    return out;
  }

  public int readByte() {
    ensure(1);
    return buffer.get() & 0xFF;
  }

  public long readUInt(int bits) {
    int bytes = bits / 8;
    ensure(bytes);
    return switch (bits) {
      case 8 -> buffer.get() & 0xFFL;
      case 16 -> buffer.getShort() & 0xFFFFL;
      case 32 -> buffer.getInt() & 0xFFFFFFFFL;
      case 64 -> buffer.getLong();
      default -> throw new IllegalArgumentException("Unsupported integer size: " + bits);
    };
  }

  public long readInt(int bits) {
    int bytes = bits / 8;
    ensure(bytes);
    return switch (bits) {
      case 8 -> buffer.get();
      case 16 -> buffer.getShort();
      case 32 -> buffer.getInt();
      case 64 -> buffer.getLong();
      default -> throw new IllegalArgumentException("Unsupported integer size: " + bits);
    };
  }

  public long readLength(boolean compact) {
    if (compact) {
      return readVarint();
    }
    return readUInt(64);
  }

  public long readVarint() {
    Varint.DecodeResult result = Varint.decode(buffer);
    return result.value();
  }

  private void ensure(int count) {
    if (buffer.remaining() < count) {
      throw new IllegalArgumentException("Unexpected end of data");
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/norito/NoritoDecoder.kt
// Copyright 2024 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.sdk.norito

import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Reader for Norito payloads. */
class NoritoDecoder(
    private val buffer: ByteBuffer,
    @JvmField val flags: Int,
) {
    constructor(data: ByteArray, flags: Int) : this(
        buffer = ByteBuffer.wrap(data.copyOf()).order(ByteOrder.LITTLE_ENDIAN),
        flags = flags,
    )

    fun compactLenActive(): Boolean = (flags and NoritoHeader.COMPACT_LEN) != 0

    fun remaining(): Int = buffer.remaining()

    fun readBytes(count: Int): ByteArray {
        ensure(count)
        val out = ByteArray(count)
        buffer.get(out)
        return out
    }

    fun readByte(): Int {
        ensure(1)
        return buffer.get().toInt() and 0xFF
    }

    fun readUInt(bits: Int): Long {
        ensure(bits / 8)
        return when (bits) {
            8 -> buffer.get().toLong() and 0xFFL
            16 -> buffer.getShort().toLong() and 0xFFFFL
            32 -> buffer.getInt().toLong() and 0xFFFFFFFFL
            64 -> buffer.getLong()
            else -> throw IllegalArgumentException("Unsupported integer size: $bits")
        }
    }

    fun readInt(bits: Int): Long {
        ensure(bits / 8)
        return when (bits) {
            8 -> buffer.get().toLong()
            16 -> buffer.getShort().toLong()
            32 -> buffer.getInt().toLong()
            64 -> buffer.getLong()
            else -> throw IllegalArgumentException("Unsupported integer size: $bits")
        }
    }

    fun readLength(compact: Boolean): Long {
        val length = if (compact) readVarint() else readUInt(64)
        require(length in 0..Int.MAX_VALUE.toLong()) {
            "Norito length exceeds the supported JVM range"
        }
        return length
    }

    fun readVarint(): Long {
        val result = Varint.decode(buffer)
        return result.value
    }

    private fun ensure(count: Int) {
        require(buffer.remaining() >= count) { "Unexpected end of data" }
    }
}
